src/components/shared/page.tsx

```diff
@@ -83,6 +85,27 @@ interface BodyProps {
 interface ElementWithResources {
     element: React.ReactElement;
     resources: string[];
+    hydrationProps: {};
+}
+
+export interface ImageMappings {
+    [key: string]: string;
+}
+
+const getImageMappings = (imageSalt: string, capi: Content): ImageMappings => {
+    return capi.blocks?.body
+        ?.flatMap(block => block.elements
+            .filter(element => element.type === ElementType.IMAGE)
+            .flatMap(element => {
+                return element.assets.map(asset => {
```

Comment:
Can this use *just* the master asset?

Reply:
yep, good idea 👍

Reply:
I *think* if you're just working with master assets you can sidestep the `map` and the second `flatMap`:

```ts
const bodyBlocks = (capi: Content): Block[] =>
    capi.blocks?.body ?? [];

const mainBlocks = (capi: Content): Block[] =>
    capi.blocks?.main !== undefined ? [capi.blocks.main] : [];

const getImageMappings = (imageSalt: string, capi: Content): ImageMappings => {
    const blocks = [...bodyBlocks(capi), ...mainBlocks(capi)];

    return blocks.flatMap(block => block.elements)
        .filter((element: BlockElement) => element.type === ElementType.IMAGE)
        .reduce((mappings, elem) => {
            const asset = elem.assets.find(asset => asset.typeData?.isMaster);

            if (asset !== undefined && asset.file !== undefined) {
                const url = new URL(asset.file);
                return { ...mappings, [url.pathname]: sign(imageSalt, url.pathname) };
            }

            return mappings;
        }, {});
}
```

Reply:
I don't think there's a need to change everything around as I did in the example above. However, I think the `flat` method returning `any` is something that could be remedied to increase our type safety.